pkg/components/prometheus-operator/component_conversion_test.go

```diff
@@ -0,0 +1,137 @@
+package prometheus //nolint:testpackage
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/hcl/v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	yamlserializer "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/client-go/util/jsonpath"
+
+	"github.com/kinvolk/lokomotive/pkg/components/util"
+)
+
+func renderManifests(t *testing.T, configHCL string) map[string]string {
+	name := "prometheus-operator"
+
+	component := newComponent()
+
+	body, diagnostics := util.GetComponentBody(configHCL, name)
+	if diagnostics.HasErrors() {
+		t.Fatalf("Getting component body: %v", diagnostics.Errs())
+	}
+
+	diagnostics = component.LoadConfig(body, &hcl.EvalContext{})
+	if diagnostics.HasErrors() {
+		t.Fatalf("Valid config should not return an error, got: %v", diagnostics)
+	}
+
+	ret, err := component.RenderManifests()
+	if err != nil {
+		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
+	}
+
+	return ret
+}
+
+//nolint:funlen
+func TestConversion(t *testing.T) {
+	testCases := []struct {
+		Name                   string
+		InputConfig            string
+		ExpectedConfigFileName string
+		Expected               reflect.Value
+		JSONPath               string
+	}{
+		{
+			Name:                   "verify foldersFromFilesStructure in configmap",
+			InputConfig:            `component "prometheus-operator" {}`,
+			ExpectedConfigFileName: "prometheus-operator/charts/grafana/templates/configmap-dashboard-provider.yaml",
+			Expected: reflect.ValueOf(`apiVersion: 1
+providers:
+- name: 'sidecarProvider'
+  orgId: 1
+  folder: ''
+  type: file
+  disableDeletion: false
+  allowUiUpdates: false
+  options:
+    foldersFromFilesStructure: true
+    path: /tmp/dashboards`),
+			JSONPath: "{.data.provider\\.yaml}",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
```

Comment:
This code should be de-duplicated with #964.

Reply:
How?

Reply:
https://github.com/kinvolk/lokomotive/pull/964#discussion_r492668859

Reply:
Test cases for a test file should reside within that test file. If you're talking about making some helper function generic, I would say we should only do that if it actually makes reading and maintaining the code easier (and that is assuming we even need a helper function).

Reply:
Yes, I'm talking about the helpers, not test cases of course. I think this code is complicated enough, that I'm worried about maintaining one copy of it, two is even worse :smile:

Reply:
Yes we should have a generic mechanism. This test code has allowed us to extract any parameter from a yaml file if we provide it json path.

Other tests can use this mechanism to verify if the hcl attribute is converted to corresponding yaml format.